### ISSUE TITLE
How to properly get detected handedness (Python)

### ISSUE THREAD

**Issue by ZodiacFRA** [google/mediapipe#1374]
Hi,

In the [sample code](https://google.github.io/mediapipe/solutions/hands.html#python) the handedness is printed this way: 
`  print('handedness:', results.multi_handedness)`
however this returns a string, and I cannot find a way to access the displayed fields: index, score and label, any tips?

`results.multi_handedness` is a `mediapipe.framework.formats.classification_pb2.ClassificationList` object, which is not iterable? 
 
Thanks! :)

**Comment by TigerHix** [comment]
You can enumerate through the list:

```
        for idx, hand_handedness in enumerate(hand_results.multi_handedness):
            handedness_dict = MessageToDict(hand_handedness)
            print(handedness_dict)
```

**Comment by ZodiacFRA** [comment]
Thanks a lot :) 
There is no python doc on this repo right? (I could submit a PR to improve it)

Just a side question, is the python module just a API to bind to the installed mediapipe? Or is it supposed to work on its own?

Thanks again :+1:

**Comment by TigerHix** [comment]
Actually I'm not a maintainer of the repo :) I just happened to use mediapipe in Python and figured these things out.

Yes, I believe there is no API documentation for the Python runtime. They do heavily comment the API functions in the source code.

I believe the mediapipe Python package works on its own; it doesn't rely on an installed mediapipe instance.

**Comment by jiuqiant** [comment]
MediaPipe python package is standalone and doesn't require the mediapipe repo.
The easiest way to see the API documentation and the usage examples is to call the help function.
```
>>> help(mp.solutions.hands.Hands)

class Hands(mediapipe.python.solution_base.SolutionBase)
 |  MediaPipe Hands.
 |  
 |  MediaPipe Hands processes an RGB image and returns the hand landmarks and
 |  handedness (left v.s. right hand) of each detected hand.
 |  
 |  Note that it determines handedness assuming the input image is mirrored,
 |  i.e., taken with a front-facing/selfie camera (
 |  https://en.wikipedia.org/wiki/Front-facing_camera) with images flipped
 |  horizontally. If that is not the case, use, for instance, cv2.flip(image, 1)
 |  to flip the image first for a correct handedness output.
 |  
 |  Usage examples:
 |    import cv2
 |    import mediapipe as mp
 |    mp_drawing = mp.solutions.drawing_utils
 |    mp_hands = mp.solutions.hands
 |  
 |    # For static images:
 |    hands = mp_hands.Hands(
 |        static_image_mode=True,
 |        max_num_hands=2,
 |        min_detection_confidence=0.7)
 |    for idx, file in enumerate(file_list):
 |      # Read an image, flip it around y-axis for correct handedness output (see
 |      # above).
 |      image = cv2.flip(cv2.imread(file), 1)
 |      # Convert the BGR image to RGB before processing.
 |      results = hands.process(cv2.cvtColor(image, cv2.COLOR_BGR2RGB))
 |  
 |      # Print handedness and draw hand landmarks on the image.
 |      print('handedness:', results.multi_handedness)
 |      if not results.multi_hand_landmarks:
 |        continue
 |      annotated_image = image.copy()
 |      for hand_landmarks in results.multi_hand_landmarks:
 |        print('hand_landmarks:', hand_landmarks)
 |        mp_drawing.draw_landmarks(
 |            annotated_image, hand_landmarks, mp_hands.HAND_CONNECTIONS)
 |      cv2.imwrite(
 |          '/tmp/annotated_image' + str(idx) + '.png', cv2.flip(image, 1))
 |    hands.close()
 |  
 |    # For webcam input:
 |    hands = mp_hands.Hands(
 |        min_detection_confidence=0.7, min_tracking_confidence=0.5)
 |    cap = cv2.VideoCapture(0)
 |    while cap.isOpened():
 |      success, image = cap.read()
 |      if not success:
 |        break
 |  
 |      # Flip the image horizontally for a later selfie-view display, and convert
 |      # the BGR image to RGB.
 |      image = cv2.cvtColor(cv2.flip(image, 1), cv2.COLOR_BGR2RGB)
 |      # To improve performance, optionally mark the image as not writeable to
 |      # pass by reference.
 |      image.flags.writeable = False
 |      results = hands.process(image)
 |  
 |      # Draw the hand annotations on the image.
 |      image.flags.writeable = True
 |      image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
 |      if results.multi_hand_landmarks:
 |        for hand_landmarks in results.multi_hand_landmarks:
 |          mp_drawing.draw_landmarks(
 |              image, hand_landmarks, mp_hands.HAND_CONNECTIONS)
 |      cv2.imshow('MediaPipe Hands', image)
 |      if cv2.waitKey(5) & 0xFF == 27:
 |        break
 |    hands.close()
 |    cap.release()
 |  
 |  Method resolution order:
 |      Hands
 |      mediapipe.python.solution_base.SolutionBase
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  __init__(self, static_image_mode=False, max_num_hands=2, min_detection_confidence=0.7, min_tracking_confidence=0.5)
 |      Initializes a MediaPipe Hand object.
 |      
 |      Args:
 |        static_image_mode: If set to False, the solution treats the input images
 |          as a video stream. It will try to detect hands in the first input
 |          images, and upon a successful detection further localizes the hand
 |          landmarks. In subsequent images, once all "max_num_hands" hands are
 |          detected and the corresponding hand landmarks are localized, it simply
 |          tracks those landmarks without invoking another detection until it loses
 |          track of any of the hands. This reduces latency and is ideal for
 |          processing video frames. If set to True, hand detection runs on every
 |          input image, ideal for processing a batch of static, possibly unrelated,
 |          images. Default to False.
 |        max_num_hands: Maximum number of hands to detect. Default to 2.
 |        min_detection_confidence: Minimum confidence value ([0.0, 1.0]) from the
 |          hand detection model for the detection to be considered successful.
 |          Default to 0.7.
 |        min_tracking_confidence: Minimum confidence value ([0.0, 1.0]) from the
 |          landmark-tracking model for the hand landmarks to be considered tracked
 |          successfully, or otherwise hand detection will be invoked automatically
 |          on the next input image. Setting it to a higher value can increase
 |          robustness of the solution, at the expense of a higher latency. Ignored
 |          if "static_image_mode" is True, where hand detection simply runs on
 |          every image. Default to 0.5.
 |  
 |  process(self, image:numpy.ndarray) -> <class 'NamedTuple'>
 |      Processes an RGB image and returns the hand landmarks and handedness of each detected hand.
 |      
 |      Args:
 |        image: An RGB image represented as a numpy ndarray.
 |      
 |      Raises:
 |        RuntimeError: If the underlying graph occurs any error.
 |        ValueError: If the input image is not three channel RGB.
 |      
 |      Returns:
 |        A NamedTuple object with two fields: a "multi_hand_landmarks" field that
 |        contains the hand landmarks on each detected hand and a "multi_handedness"
 |        field that contains the handedness (left v.s. right hand) of the detected
 |        hand.
 |  
 |  ----------------------------------------------------------------------
 |  Methods inherited from mediapipe.python.solution_base.SolutionBase:
 |  
 |  close(self) -> None
 |      Closes all the input sources and the graph.
 |  
 |  ----------------------------------------------------------------------
 |  Data descriptors inherited from mediapipe.python.solution_base.SolutionBase:
 |  
 |  __dict__
 |      dictionary for instance variables (if defined)
 |  
 |  __weakref__
 |      list of weak references to the object (if defined)

```


```

**Comment by matanster** [comment]
Thanks for this issue, indeed it seems not directly documented other than here, that this conversion needs to be used to extract the values in python.

**Comment by matanster** [comment]
I wonder whether there's a more api idiomatic way to get the data though

**Comment by RohitSingh1226** [comment]
I dont have a hand_handedness variable 
As I am trying to loop it says none type objects are not iterable. 
Could you explain what hand_handedness is?

**Comment by RohitSingh1226** [comment]
Thanks for the prompt @ZodiacFRA 
I was actually concerned with the for loop, hand_handedness was empty for my code case. 
That is why I was getting none type not iterable 

I fixed my error using the conversion from protobuf to python 

        if self.results.multi_handedness:
            for idx, hand_handedness in enumerate(self.results.multi_handedness):
                handedness_dict = MessageToDict(hand_handedness)
                whichHand =(handedness_dict['classification'][0]['label'])
                #print(whichHand)
                if whichHand == "Left":
                    RightHand = True
                else:
                    RightHand = False


        if self.results.multi_hand_landmarks and RightHand:
            for handLms in self.results.multi_hand_landmarks:
                if draw:
                    self.mpDraw.draw_landmarks(img, handLms,
                    self.mpHands.HAND_CONNECTIONS)
        return img

**Comment by JinYeJin** [comment]
Don't need to use MessageToDict.

```python
for idx, hand_handedness in enumerate(hand_results.multi_handedness):
    print(hand_handedness.classification)
    print(hand_handedness.classification[0].label)
```

I don't know why they make it too difficult to access!

**Comment by jsliang0417** [comment]
> Don't need to use MessageToDict.
> 
> ```python
> for idx, hand_handedness in enumerate(hand_results.multi_handedness):
>     print(hand_handedness.classification)
>     print(hand_handedness.classification[0].label)
> ```
> 
> I don't know why they make it too difficult to access!

Very great solution! Still don't know why so hard to access it lol.

**Comment by matanster** [comment]
You know ..... making data structures written in C accessible for python is a lot of extra work. So you can iterate them and that solves it. BTW you can't iterate them in a python comprehension last I recall.

**Comment by matanster** [comment]
I notice that the value returned for the `index` attribute of this struct is sometimes larger than the largest index of the landmark lists themselves. I.e. the index can be 1 when there is only one hand (one item in `mutli_hand_landmarks`, whose index is zero). So it's hopefully safe to ignore it but may warrant some clarification.

**Comment by mokadevcloud** [comment]
@matanster  I think the index refers to left (0) or right (1), and doesn't depend on the number of hands in the list.

**Comment by bernduwiesner** [comment]
A simple way to reference the fields:

if len(results.multi_handedness) == 1:
    print(results.multi_handedness[0].classification[0].index)
    print(results.multi_handedness[0].classification[0].label)
if len(results.multi_handedness) == 2:
    print(results.multi_handedness[0].classification[0].index)
    print(results.multi_handedness[0].classification[0].label)
    print(results.multi_handedness[1].classification[0].index)
    print(results.multi_handedness[1].classification[0].label)

**Comment by matanster** [comment]
@bernduwiesner so then what are the semantics of `results.multi_handedness[i].classification[j].index`? does it refer to the index under `results.multi_hand_landmarks`, of the hand that the handedness classification associates to? I have [previously seen](https://github.com/google/mediapipe/issues/1374#issuecomment-1013428464) values to suggest otherwise, but maybe I had been wrong there. 

Or does it merely repeat the index of the `results.multi_handedness` list that it is sitting under?

I would say the indexing scheme here could sure enjoy more clarity.

**Comment by bernduwiesner** [comment]
The results reflect only an instant in time (a single frame) and when one considers that at one instant there may be 7 left hands and one right hand and in the next instant there are only two right hands then it may be seen that the id of a particular hand may well change over time. For example if the first detected of two hands (0 and 1) is removed then the remaining hand changes id and becomes 0 because in that frame it is the first hand.
In other words the id is not associated with a specific hand except at one instant (frame). However as a hand is tracked and remains trackable then it may retain an id, but is not guaranteed to do so. 
At least this is what I have understood from  experimentation.
